src/Provider/DeferredProvider.php

```diff
@@ -0,0 +1,67 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\EventDispatcher\Provider;
+
+use Psr\EventDispatcher\EventDispatcherInterface;
+use Psr\EventDispatcher\ListenerProviderInterface;
+
+/**
+ * DeferredProvider is a listener provider that allows you to perform deferred
+ * events
+ */
+final class DeferredProvider implements ListenerProviderInterface
+{
+    private EventDispatcherInterface $dispatcher;
+    private bool                     $deferEvents = false;
+    private array                    $events = [];
+
+    public function __construct(EventDispatcherInterface $dispatcher)
+    {
+        $this->dispatcher = $dispatcher;
+    }
+
+    public function getListenersForEvent(object $event): iterable
+    {
+        if ($this->deferEvents) {
+            yield fn($event) => $this->addEvent($event);
+        } else {
+            yield fn($event) => $this->dispatchEvent($event);
+        }
+    }
+
+    public function deferEvents(): void
```

Comment:
Looks like w/o calling this option it works the same way as regular provider. I think it is better to add this option to main provider instead of having separate `DeferredProvider`.

Reply:
It works not only as a provider, but also as a dispatcher. Since it needs to send events, it needs an original dispatcher. Therefore, this functionality cannot be transferred to the main provider.

For good, this functionality is better implemented in the deferred dispatcher. But as we found out earlier, this will violate the psr.